src/gui/properties/speedplotview.cpp

```diff
@@ -350,31 +361,40 @@ void SpeedPlotView::paintEvent(QPaintEvent *)
     painter.setRenderHints(QPainter::Antialiasing);
 
     // draw graphs
-    rect.adjust(3, 0, 0, 0); // Need, else graphs cross left gridline
+    // averager is duration based, it may go little above the maxDuration
+    painter.setClipping(true);
+    painter.setClipRect(rect);
 
-    const double yMultiplier = (niceScale.arg == 0.0) ? 0.0 : (static_cast<double>(rect.height()) / niceScale.sizeInBytes());
-    const double xTickSize = static_cast<double>(rect.width()) / (m_viewablePointsCount - 1);
+    const DataCircularBuffer &queue = currentData();
 
-    boost::circular_buffer<PointData> &queue = getCurrentData();
+    // last point will be drawn at x=0, so we don't need it in the calculation of xTickSize
+    const milliseconds lastDuration {queue.empty() ? milliseconds {} : queue.back().duration};
+    const double xTickSize = static_cast<double>(rect.width()) / (m_currentMaxDuration - lastDuration).count();
+    const double yMultiplier = (niceScale.arg == 0.0) ? 0.0 : (static_cast<double>(rect.height()) / niceScale.sizeInBytes());
 
     for (int id = UP; id < NB_GRAPHS; ++id)
     {
         if (!m_properties[static_cast<GraphID>(id)].enable)
             continue;
 
         QVector<QPoint> points;
-        for (int i = static_cast<int>(queue.size()) - 1, j = 0; (i >= 0) && (j < m_viewablePointsCount); --i, ++j)
-        {
+        milliseconds duration {};
 
-            int newX = rect.right() - j * xTickSize;
-            int newY = rect.bottom() - queue[i].y[id] * yMultiplier;
+        for (int i = static_cast<int>(queue.size()) - 1; i >= 0; --i)
+        {
+            int newX = rect.right() - (duration.count() * xTickSize);
```

Comment:
IIRC, each period is drawn proportionally to its duration, am I right?

Reply:
yes, and note that I am drawing backwards.